api/v1alpha4/indexer.go

```diff
@@ -0,0 +1,92 @@
+/*
```

Comment:
Should we move this under `controllers/noderefutil`?

Reply:
Since the fields that are indexed are version specific the idea is to put this logic in the api package and take it out of the controller. Ref: https://github.com/kubernetes-sigs/cluster-api/pull/3748#discussion_r503548847

Reply:
fwiw I have a different perspective on this, I'd like `api/` to be as lean as possible and ideally to eventually become a module on its own so consumers are not dictated with versions of transitive deps they wouldn't need to consume the API.

Reply:
Apologies I should have reworded the issue a bit. We should keep constants and whatnot exported in the API package, but let's add the indexing logic (and the import on controller runtime) under `noderefutil`, how does it sound?

cc @ncdc

Reply:
🚀 SGTM

Reply:
👍🏼  will update the PR.